### ISSUE TITLE
Shadow DOMs block EVERYTHING! Not just each others' CSS

### ISSUE THREAD

**Issue by MarkPuchalaII** [MarkPuchalaII/Reusable-HTML-Components#1]
index.html is utilizing a reusable footer component to display its footer, which has been placed in its own shadow DOM to prevent its CSS from changing other elements on the page. 
This has, however, resulted in the footer being unable to render the social media icons brought in from the page's font-awesome link.

I've created indexWithoutShadowDOM.html for an easy image of what things _should_ look like.

I can't shake the feeling that it would be redundant to simply place an font-awesome link into the component, 
as that would result in rendering two links to the same page for the webpage that this component's being applied to.

I wonder if the question is: How do we link to an external CSS sheet on the webpage, and _not_ on a reusable component in a shadow DOM, all while ensuring the shadow DOM can still get what it needs from that same external CSS link? 

I worry that this question is just another way of wording "let the webpage end up with two links to the same sheet".

So that's why I think this is currently beyond my skills.
Any ideas?

**Comment by scissorsneedfoodtoo** [comment]
Hey @MarkPuchalaII, thanks again for reaching out about this issue and for creating this repo. Also, thank you for all the helpful details you provided.

Unfortunately I may have missed this when I was originally working on the article a year ago. The good thing is that there seem to be a few simple solutions.

The first, like you mentioned, is just to include the same Font Awesome link within the component itself. It's not the DRY-est thing in the world, but there doesn't seem to be a downside outside of that. As far as I can tell, even though you're linking to Font Awesome in both the main and Shadow DOM, FA is only loaded once.

Here's a quick screenshot of the Network tab after I included a link to the FA CDN in `components/footer.js`:

![Screenshot from 2021-09-12 14-53-01](https://user-images.githubusercontent.com/2051070/132973806-5b870c9e-2c8e-496a-97be-47bb75f6917f.png)

Another option would be to use `@import` in `components/footer.js`:

```css
  <style>
    @import url("https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.1/css/all.min.css");

    footer {
      height: 60px;
      padding: 0 10px;
      ...
```

Note that the URL needs to match the `link` in `index.html`.

Finally, you could use a bit of JS to dynamically create a `link` element and add it to `index.html` as described in this comment: https://stackoverflow.com/a/62184131

I played around with it a bit and it seems to work well. Might be a good way to play around with JS modules in the browser, too. 

But in my testing, it seems like you would still need to add the same link or use `@import` within `components/footer.js` like before, so it may not be worth the trouble for simpler projects.

Anyway, hope that helps. Thanks again for reaching out about this, and I'll work on updating the article with this information.

By the way, how are you planning to use reusable HTML components in your personal project?

Edit: I just finished [updating the article](https://www.freecodecamp.org/news/reusable-html-components-how-to-reuse-a-header-and-footer-on-a-website/) with the methods above, and a simpler JS implementation. Please take a look if you can, and let me know if I can help with anything else.

**Comment by MarkPuchalaII** [comment]
I tested them out, and I'd say option #2 has got to be the winner.
On my project: I'm working on a personal intranet, so I can have my own big interface at home to manage all my projects from, as well as practice treating everything I work on as part of some greater business portfolio.

This has led me to so many web pages that I want a good set of header and footer components to easily insert into whichever pages may need them. But my intranet is also a practice in showcasing the best of my skills, which was where option #1 felt too WET, and I just couldn't stand by it.

With option #2, one could argue "you have the same link twice," and I'll admit I wonder about some sort of conditional HTML injection, like `If !link { add link }`, but I feel that hides the more underlying issue that I think HTML5 itself could use some improvements with the accessibility of components in general.

Using option #2, it really seems to send the message like, "Yeah I know I used the link twice, but you can see it's packaged in the component in such a particular way that I know this isn't my fault, and I'm doing my due diligence to make sure people can still see that this component NEEDS its link whether the page it's on does or not".

I feel like option 3 fails that assurance, which is again why I'm going to choose #2. I notice how, amidst how clever option 3 seems, It's appending HTML tags that use `class="fab fa-github"`, but can't actually link to font-awesome if the webpage it's appended to doesn't have the link. Regardless of how small the risk, I still feel that's too much risk of the footer one day breaking when it gets added to a webpage that isn't using font-awesome.

### TL;DR:

1. repasting the same tag looks WET, whether I know better or not.
2. Sends a message: MAYBE looks wet, but at least I'm communicating what this component needs
3. might break if the page doesn't already have the tool.

So I would say option #2 is the clear winner!


P.S. I have to say, I would NOT have found #2 or #3 without you. This one had really been stumping me. 
I feel so much more competent in my work thanks to your help on this matter.

P.P.S I see this as a case closed, but I'm still not so experienced with issues on GitHub. Should I have closed with THIS comment, or confirm you saw it first?

**Comment by scissorsneedfoodtoo** [comment]
Thanks again for that detailed breakdown, @MarkPuchalaII. I agree that option #2 is a good one for the reasons you listed above.

Also, I agree with the potential pitfalls you mentioned about #3 above, and ended up using a modified version to update the article:

```js
...
class Footer extends HTMLElement {
  constructor() {
    super();
  }

  connectedCallback() {
    const fontAwesome = document.querySelector('link[href*="font-awesome"]');
    const shadowRoot = this.attachShadow({ mode: 'closed' });

    if (fontAwesome) {
      shadowRoot.appendChild(fontAwesome.cloneNode());
    }

    shadowRoot.appendChild(footerTemplate.content);
  }
}
```

The snippet above checks for a `link` element with an `href` that includes the text `font-awesome`. If it exists, it clones the `link` and appends the clone to the `shadowRoot`. If a `link` element pointing to `font-awesome` doesn't exist, then nothing should happen.

But there's no clear winner between option #2 and the code above as far as I can tell. It's just a matter of whether you'd rather mess around with CSS or with JS, and which is more explicit to you. Either method involves some extra work to get Font Awesome working in the Shadow DOM.

About the issue, you can go ahead and close it if you feel it's been solved! Better yet, you could create a PR here using your preferred method, and close this issue once that's merged.

And thank you for reaching out about this issue on Twitter. I probably wouldn't have noticed it without your help and all the details you provided.

Also, your project sounds really interesting! Hope you put up some code for it on GitHub when you're ready :)
